acp/acp.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package acp
+
+import (
+	"context"
+
+	"github.com/sourcenetwork/immutable"
+
+	"github.com/sourcenetwork/corelog"
+)
+
+var (
+	log = corelog.NewLogger("acp")
+
+	// NoACPModule is an empty ACPModule, this is used to disable access control.
+	NoACPModule = immutable.None[ACPModule]()
+)
+
+// ACPModule is the interface to all types of access control modules that might exist.
+type ACPModule interface {
+	// Init initializes the acp module, with an absolute path. The provided path indicates where the
+	// persistent data will be stored for the acp module.
+	Init(ctx context.Context, path immutable.Option[string])
+
+	// Start starts the acp module, using the initialized path. Will recover acp module state
+	// from a previous run if under the same path.
+	//
+	// If the path does not have a value, then acp module will run in memory.
+	// If the path has a value, but the value is empty then returns an error.
+	Start(ctx context.Context) error
+
+	// Close closes the resources in use by the acp module.
+	Close() error
+
+	// AddPolicy attempts to add the given policy. Detects the format of the policy automatically
+	// by assuming YAML format if JSON validation fails. Upon success a policyID is returned,
+	// otherwise returns error.
+	//
+	// A policy can not be added without a creator identity (sourcehub address).
+	AddPolicy(ctx context.Context, creatorID string, policy string) (string, error)
+
+	// ValidateResourceExistsOnValidDPI performs DPI validation of the resource (matching resource name)
+	// that is on the policy (matching policyID), returns an error upon validation failure.
+	//
+	// Learn more about the DefraDB Policy Interface [DPI](/acp/ACP.md)
+	ValidateResourceExistsOnValidDPI(
```

Comment:
> nitpick: Not the best naming here, should be more succinct. Perhaps just `ValidateDPIResource`.
> 
> Naming aside, in the future we should move this kind of validation off the interface, as the `DPI` is only one kind of structure a policy can use. Looking forward, we dont need DPI validator for Admin ACP stuff, but still needs some kind of validation. So this will need to be better abstracted.

I already plan to move this off the ACP interface in future. 

Regarding the name change I can do but as you mentioned we will need other types of validation for AAC (Admin Access Control) too, so might be nice to rename to cater to these "different" types of validation.

Eventually I want to rename it to something like: `ValidateDACResource`, and `ValidateAACResource` etc.

Keeping as is for now, unless a harder preference, or consensus.